Feature: Notes (Example first implementation)

dearshrewdwit opened the pull request:
## Notes

Standard Criteria
```
When a user navigates to a user’s repo page ‘/dearshrewdwit/bowling-challenge’, then a user should see, in addition to the information about a repo, a notes section. This includes:
- A clickable element to add a note
- All current notes listed down the page in reverse chronological order (most recent at the top) for the repository
- Notes should be persisted to a data store

When a user clicks to add a note, then the url should change to ‘/dearshrewdwit/bowling-challenge/notes/add’ and show a text input field and a button to add the note. When the note is added, the app should navigate back to the repo page and the user should see all the notes including the newly created one.
 ```
 
Extensions
```
Set 1
1. When a user navigates to the index ('/') they should only see the input field and button. 
2. When a user types in a github username, they should see a list of repositories and also some other information about the user's github profile - their avatar image at minimum. The URL should also show as '/:username'
3. A user should be able to navigate back to the user's profile page '/:username' from the repo page 
4. A user should be able to navigate back to the home page from anywhere.
```
```
Set 2
1. When a user navigates to a user’s repo page ‘/dearshrewdwit/bowling-challenge’, then a user should see, in addition to the standard criteria, notes with clickable elements to edit and delete each note
2. When a user clicks to edit a note, then the url should change to ‘/dearshrewdwit/notes/12/edit’ and show a text input field populated with the current note text and a button to update the note. When the note is updated, the app should navigate back to the repo page and the user should see all the notes including the newly updated one.
3. When a user navigates to ‘/notes’, they should see all the notes added to any user’s repository. Notes should be displayed with the note contents and the associated username and repo name
4. When a user navigates to ‘/notes’ they should see a search input field which filters the notes to only those notes whose contents includes the characters searched for. The filtering should happen automatically on change to the input field and the page should always show the current filtered list of notes.
```

dearshrewdwit committed:
Notes Feature: standard criteria

dearshrewdwit committed:
Notes Feature: Extensions - set 1

dearshrewdwit committed:
Notes Feature: Extensions 1, 2

- User can edit and delete notes
- Navigate back to repo page after edits and deletes

dearshrewdwit committed:
Notes Feature: Extensions 3, 4

- User can see all notes with contents, time, repo, owner
- User can search notes by contents

desokan commented on src/components/Repos/Repos.js:
```diff
@@ -1,56 +1,39 @@
         setNotFound(false)
-        setRepos(data)
+        setUser(data)
+        const res2 = await fetch(`https://api.github.com/users/${username}/repos`)
```
What would be the difference if we put this in a separate useEffect()?

desokan commented on src/components/Repos/Repos.js:
```diff
@@ -1,56 +1,39 @@
+    const getUserAndRepos = async () => {
+      const res = await fetch(`https://api.github.com/users/${username}`)
+      const data = await res.json()
       if (data.message === 'Not Found') {
```
Where is the message parameter from?

desokan commented on src/components/Repos/Repo.js:
```diff
@@ -1,45 +1,72 @@
-      .then(res => res.json()) // read the response format which is stored in JSON
-      .then(data => {
-      if (data.message === 'Not Found') {
+    const getRepoAndNotes = async () => {
```
I noticed that you used async-await instead of then() method. Is there any difference between them and any reason why one is preferred over the other?

procrastian commented on src/components/Repos/Repos.js:
```diff
@@ -1,56 +1,39 @@
-    fetch(`https://api.github.com/users/${username}/repos`)
-      .then(res => res.json()) // read the response format which is stored in JSON
-      .then(data => {
+    const getUserAndRepos = async () => {
```
Is using this async function with await a different way of having some kind of "if true => do this"?

ample-samples commented on src/components/Repos/Repos.js:
```diff
@@ -1,56 +1,39 @@
 function Repos () {
   const [repos, setRepos] = useState([])
-  const [username, setUsername] = useState('dearshrewdwit')
-  const [formData, setFormData] = useState(initialFormData)
+  const [user, setUser] = useState({})
   const [notFound, setNotFound] = useState(false)
+  const { username } = useParams()
```
How come you chose to initialise your state in Repos.js? Might it make more sense to do this in App.js and and pass these variables down through components?

procrastian commented on src/components/Repos/NotesEditForm.js:
```diff
@@ -0,0 +1,44 @@
+
+function NotesEditForm() {
+  const { username, noteId } = useParams()
+  const [note, setNote] = useState({contents: '', repo: ''})
```
Why is this state initialised using only contents and repo, is he fetch request (line 9) not fetching the whole object for the matching ID from your data store?

Shylan21 commented on src/components/Repos/NotesEditForm.js:
```diff
@@ -0,0 +1,44 @@
+  useEffect(() => {
+    const getNote = async () => {
+      const res = await fetch(`http://localhost:4000/notes/${noteId}`)
+      const data = await res.json()
```
Could you please explain 'async' and 'await'. I've looked them up but it's not really clear to me

procrastian commented on src/components/Repos/NotesForm.js:
```diff
@@ -0,0 +1,39 @@
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      body: JSON.stringify({...note, timestamp: new Date()})
```
Where is the ID being added in to your data store objects? Handle change is changing the content value, and this is adding the timestamp when it is posted, but I may have missed where the ID is being generated and how it is incrementing to ensure it is unique for each note being added?

alexjshaw replied:
I believe that when you create the object the ID is automatically generated.

Shylan21 commented on src/components/Repos/Repo.js:
```diff
@@ -1,45 +1,72 @@
+  }, [username, reponame])
 
+  const sortedNotes = () => {
+    return repo.notes.sort((a, b) => Date.parse(b.timestamp) - Date.parse(a.timestamp))
```
Does this changes the order of the displayed notes?? So instead of adding a new one at the bottom it gets added at the top?

procrastian commented on src/components/Repos/Repo.js:
```diff
@@ -1,45 +1,72 @@
+  }, [username, reponame])
 
+  const sortedNotes = () => {
+    return repo.notes.sort((a, b) => Date.parse(b.timestamp) - Date.parse(a.timestamp))
```
Could you explain the use of a and b parameters here please?

dearshrewdwit replied:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort

ManNavic commented on src/components/Repos/Notes.js:
```diff
@@ -0,0 +1,50 @@
+
+
+  const sortedNotes = () => {
+    return notes.sort((a, b) => Date.parse(b.timestamp) - Date.parse(a.timestamp))
```
what does the data.parse do ?

alexjshaw commented on src/components/Repos/Repo.js:
```diff
@@ -1,45 +1,72 @@
 
 function Repo () {
-  const [repo, setRepo] = useState({})
+  const [repo, setRepo] = useState({notes: []})
```
This kind of approach didn't even occur to me, using filtering on the fetch requests to build an array of all notes for the current repo. My method was to have one state for Repo, with the current Repo name stored in, and one state of Notes which contained all notes in the json, and then filter the notes array inside my map function when displaying them.

Your method is much more efficient!

dearshrewdwit committed:
Extract SearchPage

dearshrewdwit committed:
Refactor: Extract pages from components

- create pages for each page a user visits
- update components to be reusable where possible
- extract a client for requesting notes data
- use a constants file
- Inject note type component
